File: android/core/src/main/kotlin/app/fihaven/core/logic/Income.kt
package app.fihaven.core.logic

import app.fihaven.core.model.IncomeAdjustment
import app.fihaven.core.model.IncomeSource
import app.fihaven.core.model.income
import app.fihaven.core.model.incomeAdjustments
import app.fihaven.core.model.incomes
import kotlinx.serialization.json.JsonObject
import java.time.temporal.ChronoUnit

/// Income-frequency normalization, ported from income.js.
object Income {
    data class Frequency(val key: String, val label: String, val perMonth: Double)

    val frequencies: List<Frequency> = listOf(
        Frequency("hourly", "Hourly", 52.0 / 12.0), // ×hoursPerWeek
        Frequency("weekly", "Weekly", 52.0 / 12.0),
        Frequency("biweekly", "Bi-weekly", 26.0 / 12.0),
        Frequency("semimonthly", "Semi-monthly", 2.0),
        Frequency("monthly", "Monthly", 1.0),
        Frequency("annual", "Annual", 1.0 / 12.0),
    )

    /** Weeks per month — converts an hourly rate (× hours/week) to monthly. */
    const val WEEKS_PER_MONTH: Double = 52.0 / 12.0

    fun factor(frequency: String): Double =
        frequencies.firstOrNull { it.key == frequency }?.perMonth ?: 1.0

    /** Monthly equivalent of a source. Hourly = rate × hours/week × weeks/month. */
    fun monthly(source: IncomeSource): Double =
        if (source.frequency == "hourly") source.amount * source.hoursPerWeek * WEEKS_PER_MONTH
        else source.amount * factor(source.frequency)

    /// Base monthly income (recurring sources only).
    fun monthlyIncome(settings: JsonObject): Double {
        val sources = settings.incomes
        return if (sources.isNotEmpty()) sources.sumOf { monthly(it) } else settings.income
    }

    /// Adjustments (bonuses / unpaid time off / raises) affecting period [mk].
    fun adjustmentsFor(settings: JsonObject, mk: String): List<IncomeAdjustment> =
        settings.incomeAdjustments.filter { it.appliesTo(mk) }

    /// Signed total of all adjustments affecting period [mk].
    fun adjustmentsTotal(settings: JsonObject, mk: String): Double =
        adjustmentsFor(settings, mk).sumOf { it.amount }

    /// Effective income for a period: base income + applicable adjustments.
    fun monthlyIncome(settings: JsonObject, mk: String): Double =
        monthlyIncome(settings) + adjustmentsTotal(settings, mk)

    const val AVG_MONTH_DAYS: Double = 365.0 / 12.0

    fun periodDays(bounds: PeriodBounds): Int =
        ChronoUnit.DAYS.between(bounds.start, bounds.end).toInt().coerceAtLeast(1)

    data class MonthOverlap(val mk: String, val fraction: Double)

    fun monthOverlaps(bounds: PeriodBounds): List<MonthOverlap> {
        val out = mutableListOf<MonthOverlap>()
        var cursor = bounds.start.withDayOfMonth(1)
        while (cursor.isBefore(bounds.end)) {
            val monthStart = cursor
            val monthEnd = monthStart.plusMonths(1)
            val overlapStart = if (bounds.start.isAfter(monthStart)) bounds.start else monthStart
            val overlapEnd = if (bounds.end.isBefore(monthEnd)) bounds.end else monthEnd
            val overlapDays = ChronoUnit.DAYS.between(overlapStart, overlapEnd)
            val monthDays = ChronoUnit.DAYS.between(monthStart, monthEnd)
            if (overlapDays > 0 && monthDays > 0) {
                out += MonthOverlap(DateLogic.monthKey(monthStart), overlapDays.toDouble() / monthDays.toDouble())
            }
            cursor = monthEnd
        }
        return out
    }

    fun adjustmentsTotal(settings: JsonObject, bounds: PeriodBounds): Double {
        if (bounds.mode == "calendar") return adjustmentsTotal(settings, bounds.key)
        return monthOverlaps(bounds).sumOf { (mk, fraction) ->
            adjustmentsTotal(settings, mk) * fraction
        }
    }

    fun periodIncome(settings: JsonObject, bounds: PeriodBounds): Double {
        val base = monthlyIncome(settings)
        if (bounds.mode == "calendar") return base + adjustmentsTotal(settings, bounds.key)
        val prorate = periodDays(bounds) / AVG_MONTH_DAYS
        return base * prorate + adjustmentsTotal(settings, bounds)
    }

    fun incomeLabel(cfg: PeriodConfig): String =
        if (cfg.mode == "calendar") "Monthly income" else "Period income"

    fun owedLabel(cfg: PeriodConfig): String =
        if (cfg.mode == "calendar") "Left to pay" else "Left this period"
}

File: ios/FiHavenCore/Sources/FiHavenCore/Logic/Income.swift
import Foundation

/// Income-frequency normalization, ported from income.js. All three
/// clients must agree on these factors so the runway/budget numbers match.
public enum Income {
    public struct Frequency: Equatable, Sendable {
        public let key: String
        public let label: String
        public let perMonth: Double
    }

    public static let frequencies: [Frequency] = [
        Frequency(key: "hourly",      label: "Hourly",       perMonth: 52.0 / 12.0), // ×hoursPerWeek
        Frequency(key: "weekly",      label: "Weekly",       perMonth: 52.0 / 12.0),
        Frequency(key: "biweekly",    label: "Bi-weekly",    perMonth: 26.0 / 12.0),
        Frequency(key: "semimonthly", label: "Semi-monthly", perMonth: 2),
        Frequency(key: "monthly",     label: "Monthly",      perMonth: 1),
        Frequency(key: "annual",      label: "Annual",       perMonth: 1.0 / 12.0),
    ]

    /// Weeks per month — converts an hourly rate (× hours/week) to monthly.
    public static let weeksPerMonth: Double = 52.0 / 12.0

    /// Per-month multiplier for a frequency key; unknown keys → monthly (1).
    public static func factor(for frequency: String) -> Double {
        frequencies.first { $0.key == frequency }?.perMonth ?? 1
    }

    /// Monthly equivalent of a single income source. Hourly multiplies the
    /// rate by hours/week by weeks-per-month; others by the frequency factor.
    public static func monthly(of source: IncomeSource) -> Double {
        if source.frequency == "hourly" {
            return source.amount * source.hoursPerWeek * weeksPerMonth
        }
        return source.amount * factor(for: source.frequency)
    }

    /// The user's monthly income: sum of `settings.incomes`, falling back
    /// to the legacy single `settings.income` when the list is empty.
    public static func monthlyIncome(from settings: Settings) -> Double {
        let sources = settings.incomes
        if !sources.isEmpty {
            return sources.reduce(0) { $0 + monthly(of: $1) }
        }
        return settings.income
    }

    /// Adjustments (bonuses / unpaid time off / raises) affecting period `mk`.
    public static func adjustments(from settings: Settings, monthKey mk: String) -> [IncomeAdjustment] {
        settings.incomeAdjustments.filter { $0.applies(to: mk) }
    }

    /// Signed total of all adjustments affecting period `mk`.
    public static func adjustmentsTotal(from settings: Settings, monthKey mk: String) -> Double {
        adjustments(from: settings, monthKey: mk).reduce(0) { $0 + $1.amount }
    }

    /// Effective income for a specific period: base income + applicable adjustments.
    public static func monthlyIncome(from settings: Settings, monthKey mk: String) -> Double {
        monthlyIncome(from: settings) + adjustmentsTotal(from: settings, monthKey: mk)
    }

    /// Average calendar month length — used to prorate income for non-calendar periods.
    public static let avgMonthDays: Double = 365.0 / 12.0

    /// Whole days in a period (start inclusive, end exclusive).
    public static func periodDays(_ bounds: PeriodBounds, tz: TimeZone) -> Int {
        let cal = DateLogic.calendar(tz: tz)
        return cal.dateComponents([.day], from: bounds.startDate, to: bounds.endDate).day ?? Int(avgMonthDays)
    }

    /// Calendar months overlapping a period, with the fraction of each month covered.
    public static func monthOverlaps(_ bounds: PeriodBounds, tz: TimeZone) -> [(mk: String, fraction: Double)] {
        let cal = DateLogic.calendar(tz: tz)
        var out: [(String, Double)] = []
        var cursor = cal.date(from: cal.dateComponents([.year, .month], from: bounds.startDate)) ?? bounds.startDate
        while cursor < bounds.endDate {
            let monthStart = cursor
            guard let monthEnd = cal.date(byAdding: .month, value: 1, to: monthStart) else { break }
            let overlapStart = max(bounds.startDate, monthStart)
            let overlapEnd = min(bounds.endDate, monthEnd)
            let overlapDays = cal.dateComponents([.day], from: overlapStart, to: overlapEnd).day ?? 0
            let monthDays = cal.dateComponents([.day], from: monthStart, to: monthEnd).day ?? 30
            if overlapDays > 0, monthDays > 0 {
                let c = cal.dateComponents([.year, .month], from: monthStart)
                let mk = String(format: "%04d-%02d", c.year ?? 0, c.month ?? 0)
                out.append((mk, Double(overlapDays) / Double(monthDays)))
            }
            cursor = monthEnd
        }
        return out
    }

    /// Signed total of adjustments affecting a period (prorated for non-calendar modes).
    public static func adjustmentsTotal(from settings: Settings, bounds: PeriodBounds, tz: TimeZone) -> Double {
        if bounds.mode == "calendar" {
            return adjustmentsTotal(from: settings, monthKey: bounds.key)
        }
        return monthOverlaps(bounds, tz: tz).reduce(0) { sum, pair in
            sum + adjustmentsTotal(from: settings, monthKey: pair.mk) * pair.fraction
        }
    }

    /// Effective income for the active budgeting period.
    public static func periodIncome(from settings: Settings, bounds: PeriodBounds, tz: TimeZone) -> Double {
        let base = monthlyIncome(from: settings)
        if bounds.mode == "calendar" {
            return base + adjustmentsTotal(from: settings, monthKey: bounds.key)
        }
        let days = Double(periodDays(bounds, tz: tz))
        let prorate = days / avgMonthDays
        return base * prorate + adjustmentsTotal(from: settings, bounds: bounds, tz: tz)
    }

    public static func incomeLabel(for config: PeriodConfig) -> String {
        config.mode == "calendar" ? "Monthly income" : "Period income"
    }

    public static func owedLabel(for config: PeriodConfig) -> String {
        config.mode == "calendar" ? "Left to pay" : "Left this period"
    }
}
